### PR TITLE
Refactor velocity model

### PR DESCRIPTION
**Pull request by alexanderkuvaev**
An interface for stacking velocity picking was unified with the other methods of `SeismicPro`:
* Now it accepts `init` and `bounds` similar to `RefractorVelocity.from_first_breaks`
  * This allows for more precise definition of area for velocity picking: trapezoid used before was inadequate for velocity picking at later times
  * Velocity refinement is now possible: a field can first be roughly estimated, then smoothed and, finally, passed as `init` with a smaller `relative_margin` to fine-tune it and utilize extra spatial information
* `n_times` and `n_velocities` were changed to their absolute-valued steps analogues
  * We usually work with SEG-Y files cut in time and this required fine-tuning `n_times` for each field. Now the default value of 100ms step should generally work well.
  * Previously uniform spacing of velocities produced a non-uniform spacing of hodograph correction times which actually matters. This is now reversed: node velocities are now selected so that hodograph times at `max_offset` are uniform with a step of `hodograph_correction_step`.
* Also, some extra features such as optional skips of times were added.

The new algorithm works faster if the graph has the same size due to internal optimizations and move from `networkx` to `rustworkx`. However, the new default parameters produce a much larger graph and thus the total picking time is increased (but still much faster that spectrum calculation though).

Velocity spectrum calculation parameters were also updated: hardcode was removed and now some parameters such as `stacking_velocity` and `relative_margin` are passed directly to stacking velocity picking algorithm as defaults. This allows running the picking pipeline almost without any extra arguments, see an example in the `README.md`.

### CODE REVIEW

**Review by anton-br — APPROVED** [review]

File: seismicpro/stacking_velocity/velocity_model.py [hunk 1]
```diff
@@ -1,201 +1,208 @@
...
+    min_spectrum_velocity = spectrum_velocities.min()
+    max_spectrum_velocity = spectrum_velocities.max()
+    min_vel_bound = np.clip(min_vel_bound, min_spectrum_velocity, max_spectrum_velocity)
+    max_vel_bound = np.clip(max_vel_bound, min_spectrum_velocity, max_spectrum_velocity)
+    min_vel_bound = np.minimum(min_vel_bound, max_vel_bound)
+    max_vel_bound = np.maximum(min_vel_bound, max_vel_bound)
+
+    # Estimate node velocities for each time
+    final_hodograph_time_max = np.sqrt(node_times**2 + (max_offset * 1000 / min_vel_bound)**2)  # ms
+    final_hodograph_time_min = np.sqrt(node_times**2 + (max_offset * 1000 / max_vel_bound)**2)  # ms
```
Comment (nikita-klsh) [inline seismicpro/stacking_velocity/velocity_model.py:159]: please unify `max`/`min` notation for this variables
```suggestion
    max_correction_time = np.sqrt(node_times**2 + (max_offset * 1000 / min_vel_bound)**2)  # ms
    min_correction_time = np.sqrt(node_times**2 + (max_offset * 1000 / max_vel_bound)**2)  # ms
```

File: seismicpro/stacking_velocity/velocity_model.py [hunk 2]
```diff
@@ -1,201 +1,208 @@
...
+    spectrum_velocity_indices = np.arange(len(spectrum_velocities))
+    node_velocities = []
+    node_velocities_ix = []
+    for time, start, end in zip(node_times, final_hodograph_time_min, final_hodograph_time_max):
+        n_vels = int((end - start) // hodograph_correction_step) + 1
+        vels = np.sqrt(max_offset**2 * 1000**2 / (np.linspace(start, end, n_vels)[::-1]**2 - time**2))
+        vels_ix = np.interp(vels, spectrum_velocities, spectrum_velocity_indices)
+        node_velocities.append(vels)
+        node_velocities_ix.append(vels_ix)
+    node_velocities = tuple(node_velocities)
```
Comment (nikita-klsh) [inline seismicpro/stacking_velocity/velocity_model.py:169]: it appeared this tuples can not be too large

File: seismicpro/stacking_velocity/stacking_velocity.py [hunk 3]
```diff
@@ -126,6 +127,67 @@ def from_constant_velocity(cls, velocity, coords=None):
...
+            Left and right bounds of an area for stacking velocity picking. If not given, `init` must be passed.
+        relative_margin : positive float, optional, defaults to 0.2
+            A fraction of stacking velocities defined by `init` used to estimate `bounds` if they are not given.
+        acceleration_bounds : tuple of two positive floats or "auto" or None, optional
+            Minimal and maximal acceleration allowed for the stacking velocity function. If "auto", equals to the range
+            of accelerations of stacking velocities in `bounds` extended by 50% in both directions. If `None`, only
+            ensures that picked stacking velocity is monotonically increasing. Measured in meters/seconds^2.
+        times_step : float, optional, defaults to 100
+            A difference between two adjacent times defining graph nodes.
+        max_offset : float, optional, defaults to 5000
```
[the remark this answers is not in the document]
Comment (alexanderkuvaev) [reply seismicpro/stacking_velocity/stacking_velocity.py:156]: You want hodograph correction step to be the same for all gathers in a survey and even for different surveys.

File: seismicpro/stacking_velocity/velocity_model.py [hunk 4]
```diff
@@ -1,201 +1,208 @@
...
-                                                end_velocity_range, max_vel_step, n_times, n_velocities)
-    graph = nx.DiGraph()
-    graph.add_weighted_edges_from(zip(*edges))
-    paths = nx.shortest_path(graph, source=start_node, weight="weight")  # pylint: disable=unexpected-keyword-arg
-
-    # Select only paths to the nodes at the last timestamp and choose the optimal one
-    path_weights = [(paths[end_node], nx.path_weight(graph, paths[end_node], weight="weight"))
-                    for end_node in end_nodes if end_node in paths]
-    if not path_weights:
+    spectrum_data = spectrum.velocity_spectrum.max() - spectrum.velocity_spectrum
```
[the remark this answers is not in the document]
Comment (alexanderkuvaev) [reply seismicpro/stacking_velocity/velocity_model.py:132]: I'd keep extraction of all spectrum attributes together

File: seismicpro/stacking_velocity/stacking_velocity.py [hunk 5]
```diff
@@ -126,6 +127,67 @@ def from_constant_velocity(cls, velocity, coords=None):
...
+            An offset for hodograph time estimation. Used to create graph nodes and calculate their velocities for each
+            time.
+        hodograph_correction_step : float, optional, defaults to 25
+            The maximum difference in arrival time of two hodographs starting at the same zero-offset time and two
+            adjacent velocities at `max_offset`. Used to create graph nodes and calculate their velocities for each
+            time.
+        max_n_skips : int, optional, defaults to 2
+            Defines the maximum number of subsequent times to skip to still be able to connect two nodes of the graph
+            with an edge.
```
[the remark this answers is not in the document]
Comment (alexanderkuvaev) [reply seismicpro/stacking_velocity/stacking_velocity.py:165]: Rewritten

File: seismicpro/stacking_velocity/velocity_model.py [hunk 6]
```diff
@@ -1,201 +1,208 @@
...
+    for time, start, end in zip(node_times, final_hodograph_time_min, final_hodograph_time_max):
+        n_vels = int((end - start) // hodograph_correction_step) + 1
+        vels = np.sqrt(max_offset**2 * 1000**2 / (np.linspace(start, end, n_vels)[::-1]**2 - time**2))
+        vels_ix = np.interp(vels, spectrum_velocities, spectrum_velocity_indices)
+        node_velocities.append(vels)
+        node_velocities_ix.append(vels_ix)
+    node_velocities = tuple(node_velocities)
+    node_velocities_ix = tuple(node_velocities_ix)
+
+    # Calculate allowed acceleration bounds
```
[the remark this answers is not in the document]
Comment (alexanderkuvaev) [reply seismicpro/stacking_velocity/velocity_model.py:172]: Done

File: seismicpro/stacking_velocity/velocity_model.py (hunk 1 above)
Comment (alexanderkuvaev) [reply seismicpro/stacking_velocity/velocity_model.py:159]: Done

File: seismicpro/stacking_velocity/velocity_model.py [hunk 7]
```diff
@@ -1,201 +1,208 @@
...
+        max_acceleration = max(min_vel_accelerations.max(), max_vel_accelerations.max())
+        acceleration_bounds = [min_acceleration * 0.5, max_acceleration * 1.5]
+    acceleration_bounds = np.array(acceleration_bounds, dtype=np.float32)
+    if len(acceleration_bounds) != 2:
+        raise ValueError("acceleration_bounds must be an array-like with 2 elements")
+    if acceleration_bounds[1] <= acceleration_bounds[0]:
+        raise ValueError("Upper acceleration bound must greater than the lower one")
+
+    # Create a graph and find the path with maximal velocity spectrum sum along it
+    n_nodes = 2 + sum(len(node_vels) for node_vels in node_velocities)
```
[the remark this answers is not in the document]
Comment (alexanderkuvaev) [reply seismicpro/stacking_velocity/velocity_model.py:189]: Renamed lots of vars in code

File: seismicpro/stacking_velocity/velocity_model.py (hunk 2 above)
Comment (alexanderkuvaev) [reply seismicpro/stacking_velocity/velocity_model.py:169]: Fixed

File: seismicpro/stacking_velocity/velocity_model.py [hunk 8]
```diff
@@ -1,201 +1,208 @@
...
+        vels_ix = np.interp(vels, spectrum_velocities, spectrum_velocity_indices)
+        node_velocities.append(vels)
+        node_velocities_ix.append(vels_ix)
+    node_velocities = tuple(node_velocities)
+    node_velocities_ix = tuple(node_velocities_ix)
+
+    # Calculate allowed acceleration bounds
+    if acceleration_bounds is None:
+        acceleration_bounds = [0, np.inf]
+    elif acceleration_bounds == "auto":
```
[the remark this answers is not in the document]
Comment (alexanderkuvaev) [reply seismicpro/stacking_velocity/velocity_model.py:175]: I want to allow for negative accelerations in this case. Fixed the case when min acceleration was negative and max - positive

**Review by nikita-klsh — APPROVED** [review]

**Pull request merged**